errors.md

```diff
@@ -0,0 +1,448 @@
+// is not permitted.
+func Resume() error {
+  if unauthorized {
+    return &AuthorizationError{"Resume"}
+  }
+  // Rest omitted.
+}
+```
+
+## Identifier Visibility
+Like Java, Go has a concept of identifier visibility: exported and unexported.
+Take care to ensure that your API's callers can access your error (sentinel)
+values and types as necessary.
+
+```go
+// Package purchaserecord provides a high-level datastore around user purchase
+// events. Integrators should prefer using this package than directly accessing
+// the database without this API.
+package purchaserecord
+
+type DB struct{ /* rest omitted */ }
+
+type LedgerInconsistencyError struct{
+  Event *PurchaseEvent
+}
+
+// LoadRecords loads and summarizes the account's purchase events. It returns a
+// *LedgerInconsistencyError if it detects a bookkeeping problem.
+func (db *DB) LoadRecords(accountId int) (*Summary, error) { /* rest omitted */ }
+```
+
+Suppose instead of `LedgerInconsistencyError` the package had made it
+`ledgerInconsistencyError`, where the error type is unexported (thusly
+"private" from integrators' standpoing)? Callers could not handle this
+programmatically.
+
+## Checked Exceptions
+Go has no conception of checked exceptions. It is unclear whether anything more
+needs to be said about that.
+
+## Runtime Exceptions, Panics, and Broken Invariants
+### Removing or Reducing Possibility for Errors
+It is best to design your API such that certain classes of problems are solved
+statically.  This prevents the need for entrance precondition checking.  For
+example:
+
+```java
+class Auditor {
+  // Disable user instantiation.
+  private Auditor() {}
+
+  public void record(String userName, String operation) { /* rest omitted */ }
+  
+  // Creates an Auditor that emits to the provided stream.
+  public static Auditor onStream(OutputStream stream) { /* rest omitted */ }
+  
+  // Creates an Auditor that does nothing.
+  public static Auditor noop() { /* rest omitted */ }
+  
+  // Creates an Auditor that emits records to the usual location.
+  public static Auditor newDefault() { /* rest omitted */ }
+}
+```
+
+```go
+package auditor
+
+type Auditor struct{ /* rest omitted */ }
+
+func (auditor *Auditor) Record(userName, operation string) { /* rest omitted */ }
+
+// As an analogue for "newDefault":
+//
+// Choice 1: Give Auditor a useful zero value implementation.  User creates a new auditor as such:
+//
+// var auditor Auditor
+//
+// Choice 2: Create a construction function (usually if zero value does not suffice).
+
+// New creates a new Auditor using the default configuration.
+func New() *Auditor { /* rest omitted */ }
+
+// NewWriter creates a new Auditor using the provided writer.
+func NewWriter(w io.Writer) *Auditor { /* rest omitted */ }
+
+// Noop is an auditor that does nothing.
+var Noop = NewWriter(ioutil.DiscardWriter)
+```
+
+With these specific APIs, numerous input precondition validation can be removed.
+
+### When to Panic
+
```

Comment:
"Just don't". 😉 

(I realize that's not really the answer, but it's a good default—it forces the reader to come equipped with a counterexample rather than a hypothesis).

Reply:
That may be fair advice.